azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java

```diff
@@ -283,6 +287,31 @@ public void setExecutorActive(final boolean isActive, final String host, final i
           "Set active action ignored. Executor is already " + (isActive ? "active" : "inactive"));
     }
     this.active = isActive;
+    if (!this.active) {
+      // When deactivating this executor, this call will return until every thread in {@link
+      // #createFlowRunner} to finish. When deploying new executor, old running executor will be
+      // deactivated before new one is activated and only one executor is allowed to
+      // delete/hard-linking project dirs to avoid race condition described in {@link
+      // FlowPreparer#setup}. So to make deactivation process block until flow preparation work
+      // finishes guarantees the old executor won't access {@link FlowPreparer#setup} after
+      // deactivation.
+      waitUntilFlowPreparationFinish();
+    }
+  }
+
+  /**
+   * Wait until ongoing flow preparation work finishes.
+   */
+  private void waitUntilFlowPreparationFinish() {
+    while (this.preparingFlowCount.intValue() != 0) {
```

Comment:
With the new dispatch logic, in pollExecution(), active is checked before selectAndUpdateExecution() and then submitFlow() is called. submitFlow() increments preparingFlowCount, so there could be some lag, and in this amount of time, it's possible that there could be a call to inactivate the executor, and preparingFlowCount would be zero.

With the original dispatch logic, I wasn't able to find a check for if the executor is active before the flow is submitted from ExecutorServlet.handleAjaxExecute().

Reply:
One option is to add an additional check to see if it is active after incrementing preparingFlowCount. If inactive, then skip execution and decrement preparingFlowCount. Might need to update the row in the db, so that it can be picked up by another executor for the new dispatch logic, or return an error, or some status saying the flow was not executed, with the original dispatch logic.